### ISSUE TITLE
CERTIFICATE_VERIFY_FAILED

### ISSUE THREAD

**Issue by fncap** [781flyingdutchman/background_downloader#94]
**Describe the bug**
During development, I'm using a local server with self-signed SSL certificate.
When trying to download a file from this server, following exception has thrown:

 ```
[ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: Exception: HandshakeException: Handshake error in client (OS Error: 
	CERTIFICATE_VERIFY_FAILED: application verification failure(handshake.cc:393))
```

I tried to solve this applying the known workaround as explained here https://stackoverflow.com/a/61312927 without a success.

**To Reproduce**
Try to download a file from a server with SSL self-signed certificate

**Expected behavior**
We could have different scenarios:
  - still throw an exception, in this way developer should be informed on this bad state and perform some tweakings like followings 
  - allow the linked workaround above to correctly do it's job  
  - set a specific parameter that bypass SSL verification
  - any other ideas?

Thanks!

**Comment by 781flyingdutchman** [comment]
Hi, thanks for reaching out. Does this error occur in response to a specific call to the FileDownloader? If so, can you provide details on that call, and also on what platform (eg Android) this occurs?

The issue is likely hat the http client used is a native one (in Android and iOS), not the Dart one, because the download task runs in native code. That's why the fix you refer to doesn't work, and I'm not sure there's an easy fix for this.

**Comment by fncap** [comment]
Further investigation on this issue, showed me a possible cause.
`DesktopDownloader` has `httpClient` as a static field initialised with `http.Client()` make sense ... but in this way the ` HttpOverrides.global` doesn't affect it, because it is a static field and it is valorised at compile-time!
Make sense  for you?

**Comment by 781flyingdutchman** [comment]
Yes, I am aware, but I think the bigger issue is going to be getting this to work on Android and iOS. So I think a proper solution would implement an option to configure the downloader such that it allows for this somehow. Can you check how this works for you on iOS and Android, to see if you run into similar problems using your local development server?

**Comment by 781flyingdutchman** [comment]
Looking into this a bit, it looks like you can do this for Android using a special config for your debug version, see [here](https://developer.android.com/training/articles/security-config#TrustingDebugCa). I also found an [article](https://medium.com/@noumaan/ssl-app-dev-a2923d5113c6) that appears to describe how to do this for iOS.
It's actually not clear how I can get this done in Dart, as the client needs to live in an isolate, and there is no obvious way to pass a pre-configured client to an isolate - which is why it uses a default that is therefore effectively not changeable. 

If you or someone else can figure out how to make this work, then please let me know. I'm closing this issue as I currently see no way to implement this - sorry!

**Comment by 781flyingdutchman** [comment]
Hi, I've implemented this as part of work on configuration, for Desktop and Android only.  Can you help test this?  You need to change your pubsepc.yaml to:
```
background_downloader:
    git:
      url: https://github.com/781flyingdutchman/background_downloader.git
      ref: certificate
```

To configure the downloader to bypass the certificate validation, use:
```
    final configResult = await FileDownloader()
        .configure(globalConfig: ('bypassTLSCertificateValidation', true));
```

Note the argument to `.configure` is a record (and can be a list of records), hence the brackets. The `configResult` is a list containing records where the first item is the configuration string used, followed by an empty string if all went well.  Please read the `CONFIG.md` file on the `certificate` branch to better. understand how configuration works (not published yet). A few more examples of configuration are in the current example app `main.dart`.

A few notes:
* The bypass can only be done in debug mode, not in release
* On Android, you can only turn the bypass on, not off (by setting `false` as the second parameter) and the bypass _may_ not work well for tasks that are fully in the background (i.e. run independent of the app - this doesn't happen much).

I cannot really test this functionality, so would really appreciate your help.

**Comment by 781flyingdutchman** [comment]
Hi @fncap can you perhaps take a look at this and let me know if it works as expected?  Thanks!

**Comment by 781flyingdutchman** [comment]
Included in V7.9.0, untested

**Comment by fncap** [comment]
Hi!I’ll give you a feedback after August 15th I’m out of the office now!--------------------------Francesco CapaldiInviato da iPhone 7Il giorno 5 ago 2023, alle ore 08:28, 781flyingdutchman ***@***.***> ha scritto:﻿
Hi @fncap can you perhaps take a look at this and let me know if it works as expected?  Thanks!
